src/rfq/protocols/hashflow/client.rs

```diff
@@ -0,0 +1,535 @@
+    Bytes,
+};
+
+use crate::{
+    rfq::{
+        client::RFQClient,
+        errors::RFQError,
+        models::TimestampHeader,
+        protocols::hashflow::models::{
+            HashflowMarketMakerLevel, HashflowMarketMakersResponse, HashflowPriceLevelsResponse,
+        },
+    },
+    tycho_client::feed::synchronizer::{ComponentWithState, Snapshot, StateSyncMessage},
+    tycho_common::dto::{ProtocolComponent, ResponseProtocolState},
+};
+
+#[derive(Clone)]
+pub struct HashflowClient {
+    chain: Chain,
+    chain_id: u32,
+    price_levels_endpoint: String,
+    market_makers_endpoint: String,
+    quote_endpoint: String,
+    // Tokens that we want prices for
+    tokens: HashSet<Bytes>,
+    // Min tvl value in the quote token.
+    tvl: f64,
+    http_client: Client,
+    auth_key: String,
+    source: String,
+    // Quote tokens to normalize to for TVL purposes. Should have the same prices.
+    quote_tokens: HashSet<Bytes>,
+    market_makers: Vec<String>,
+}
+
+impl HashflowClient {
+    pub fn new(
+        chain: Chain,
+        tokens: HashSet<Bytes>,
+        tvl: f64,
+        quote_tokens: HashSet<Bytes>,
+        auth_key: String,
+    ) -> Result<Self, RFQError> {
+        let chain_id = chain.id() as u32;
+
+        Ok(Self {
+            chain,
+            chain_id,
+            price_levels_endpoint: "https://api.hashflow.com/taker/v3/price-levels".to_string(),
+            market_makers_endpoint: "https://api.hashflow.com/taker/v3/market-makers".to_string(),
+            quote_endpoint: "https://api.hashflow.com/taker/v3/rfq".to_string(),
+            tokens,
+            tvl,
+            http_client: Client::new(),
+            auth_key,
+            source: "propellerheads".to_string(),
+            quote_tokens,
+            market_makers: Vec::new(), // Will be populated during first fetch
+        })
+    }
+
+    /// Normalize TVL to a common quote token for comparison
+    /// Returns the normalized TVL value, or 0.0 if normalization fails due to no liquidity
+    fn normalize_tvl(
+        &self,
+        raw_tvl: f64,
+        quote_token: Bytes,
+        levels_by_mm: &HashMap<String, Vec<HashflowMarketMakerLevel>>,
+    ) -> Result<f64, RFQError> {
+        // If the quote token is already in our approved quote token set, no conversion needed
+        if self.quote_tokens.contains(&quote_token) {
+            return Ok(raw_tvl);
+        }
+
+        // Try to find the price of the quote token in one of the approved quote tokens
+        // for normalization.
+        for approved_quote_token in &self.quote_tokens {
+            for (_mm, mm_levels_inner) in levels_by_mm.iter() {
+                for quote_mm_level in mm_levels_inner {
+                    // Check for direct pair: quote_token/approved_quote_token
+                    if quote_mm_level.pair.base_token == quote_token &&
+                        quote_mm_level.pair.quote_token == *approved_quote_token
```

Comment:
Since this is just an estimate I'm just using the first levels I find for a price estimate regardless of if there are other mms available also offering prices. Let me know if you think it's worth to average them.